.pipelines/ci.yml

```diff
@@ -25,12 +25,12 @@ variables:
     value: arosvcdev.azurecr.io/vpn
   - name: LOCAL_E2E_IMAGE
     value: arosvcdev.azurecr.io/e2e
-  - name: TAG
-    value: $(Build.BuildId)
   - name: VERSION
-    value: $(Build.BuildId)
+    value: $(TAG)
+  - name: TAG
+    value: $[ coalesce('pr-$(System.PullRequest.PullRequestId)-$(System.PullRequest.SourceCommitId)', 'master-$(Build.SourceVersion)') ]
```

Comment:
I don't think this usage of `coalesce` is going to work how you intended. Check out these docs: 

- https://learn.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#coalesce
- https://learn.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#runtime-expression-syntax

Description of `coalesce`:

> Evaluates the parameters in order (left to right), and returns the first value that doesn't equal null or empty-string.

The second doc discusses different ways to interpolate variable values into the pipeline file. Based on what's being used here, I think on master we would end up tagging the image with the literal string `pr-$(System.PullRequest.PullRequestId)-$(System.PullRequest.SourceCommitId)`.

With that said, it might take some experimenting to figure out exactly what will work.

Reply:
Thanks for pointing this out! I've updated the pipeline to ensure proper runtime evaluation of the TAG variable. The logic now uses separate variables for PR and master builds `(PR_TAG and MASTER_TAG) `and combines them with `coalesce` to dynamically construct the TAG. This resolves the issue of invalid literal strings and ensures consistent naming `(pr-[PR_ID]-[COMMIT_HASH] `for PRs and `master-[COMMIT_HASH]` for master builds).